src/hotspot/share/gc/g1/g1Policy.cpp

```diff
@@ -1088,6 +1112,18 @@ bool G1Policy::use_adaptive_young_list_length() const {
   return _young_gen_sizer.use_adaptive_young_list_length();
 }
 
+size_t G1Policy::estimate_available_eden_bytes_locked() const {
+  assert_lock_strong(Heap_lock);
+  G1Allocator* allocator = _g1h->allocator();
+  uint used = _g1h->young_regions_count();
+  uint alloc = allocator->num_nodes();
+  uint full = used - MIN2(used, alloc);
+  uint target = young_list_target_length();
+  uint remaining = target - MIN2(target, full);
+  size_t bytes = remaining * HeapRegion::GrainBytes;
+  return bytes - MIN2(bytes, allocator->used_in_alloc_regions());
```

Comment:
From what I can tell it's only this last line that require the `Heap_lock` to be held, so we could still do an estimate (slightly worse) even if we fail to take the lock. So my question is would it make sense to move the `Heap_lock->try_lock()` in here and just skip including `allocator->used_in_alloc_regions()` in the calculation when not getting the lock?

Reply:
`_g1h->young_list_count()` and `young_list_target_length()` can also be
changed, protected by `Heap_lock`.

Reply:
Also, I originally didn't include the used_in_alloc_regions information in the calculation, and the resulting predictions were pretty bad, esp. with larger region sizes.

Reply:
I see, yeah, I was mostly thinking if there was a way to make it even less likely that we would get no estimate. But if the estimate would be really bad a missing estimate might be better.